File: typescript/packages/bench/src/contracts/auditableContract.ts
/** Audit-entry production. */

/** One recorded action. */
export interface AuditEntry {
    readonly id: string;
    readonly actor: string;
    readonly action: string;
}

/** Anything that can stamp an audit entry. */
export interface AuditableContract {
    /** Who is acting, for the entry's actor field. */
    auditActor(): string;

    /** Stamp an entry with a fresh id. */
    audit(action: string): AuditEntry;
}

File: rust/crates/atelier-bench/src/contracts/auditable_contract.rs
//! Audit-entry production.

use uuid::Uuid;

/// One recorded action.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AuditEntry {
    pub id: Uuid,
    pub actor: String,
    pub action: String,
}

/// Anything that can stamp an audit entry.
pub trait AuditableContract {
    /// Who is acting, for the entry's actor field.
    fn audit_actor(&self) -> String;

    /// Stamp an entry with a fresh id.
    fn audit(&self, action: &str) -> AuditEntry {
        AuditEntry {
            id: Uuid::new_v4(),
            actor: self.audit_actor(),
            action: action.to_string(),
        }
    }
}
